tiledb/sm/rest/rest_client.cc

```diff
@@ -260,50 +260,150 @@ Status RestClient::post_query_submit(
-  while (offset < returned_data.size()) {
-    // Each serialized query object is prefixed by an 8-byte, little-endian
-    // unsigned integer that contains the total byte size of the serialized
-    // query buffer.
-    if (offset + 8 > returned_data.size()) {
-      return LOG_STATUS(
-          Status::RestError("Error fetching serialized query size."));
+  // Process all of the serialized queries contained within 'scratch'.
+  scratch->reset_offset();
+  while (scratch->offset() < scratch->size()) {
+    // We need at least 8 bytes to determine the size of the next
+    // serialized query.
+    if (scratch->offset() + 8 > scratch->size()) {
+      break;
     }
 
-    // Decode the query size.
-    const void* const prefix =
-        static_cast<uint8_t*>(returned_data.data()) + offset;
-    const uint64_t query_size = utils::endianness::decode_le<uint64_t>(prefix);
-    offset += 8;
+    // Decode the query size. We could cache this from the previous
+    // callback to prevent decoding the same prefix multiple times.
+    const uint64_t query_size =
+        utils::endianness::decode_le<uint64_t>(scratch->cur_data());
 
-    if (offset + query_size > returned_data.size()) {
-      return LOG_STATUS(Status::RestError("Error fetching serialized query."));
+    // We must have the full serialized query before attempting to
+    // deserialize it.
+    if (scratch->offset() + 8 + query_size > scratch->size()) {
+      break;
     }
 
-    // The returned data may not be 8-byte aligned. Allocate a new buffer and
-    // copy the serialized buffer into it for deserialization.
-    Buffer aux;
-    RETURN_NOT_OK(aux.write(
-        static_cast<char*>(returned_data.data()) + offset, query_size))
+    // At this point of execution, we know that we the next serialized
+    // query is entirely in 'scratch'. For convenience, we will advance
+    // the offset to point to the start of the serialized query.
+    scratch->advance_offset(8);
+
+    // We can only deserialize the query if it is 8-byte aligned. If the
+    // offset is 8-byte aligned, we can deserialize the query in-place.
+    // Otherwise, we must make a copy to an auxiliary buffer.
+    if (scratch->offset() % 8 != 0) {
+      Buffer aux;
+      st = aux.write(scratch->cur_data(), query_size);
+      assert(st.ok());
+
+      st = serialization::query_deserialize(
+          aux, serialization_type_, true, copy_state, query);
+      if (!st.ok()) {
+        scratch->set_offset(scratch->offset() - 8);
+        scratch->set_size(scratch->offset());
+        return std::max(bytes_processed, 0L);
+      }
+    } else {
+      st = serialization::query_deserialize(
+          *scratch, serialization_type_, true, copy_state, query);
+      if (!st.ok()) {
+        scratch->set_offset(scratch->offset() - 8);
+        scratch->set_size(scratch->offset());
+        return std::max(bytes_processed, 0L);
+      }
+    }
 
-    // Deserialize data returned. If the user buffers are too small to
```

Comment:
I would keep this comment in (about too small user buffers) as it is informative. Also have you tested that incomplete queries still work with this new code?

Reply:
Yeah, done locally. It'll be in the next patch.

Reply:
Done. Also I did verify that incomplete queries work with this patch. I've been testing with one of the benchmarks. I will run a few more tests and perform some byte-by-byte comparisons to verify the current behavior is maintained.